### PR TITLE
[plugin-web-app] Use locators for checkbox steps

### LINKED ISSUE #804: Add possibility to interact with checkboxes by locator
I need a possibility to interact with checkboxes by locator (not only xpaths/name) e.g., check, uncheck, verify if checkbox exists cause sometimes a checkbox does not have meaningful unique xpath.

### PR DESCRIPTION
**Pull request by ikalinin1**
closes #804

### CODE REVIEW

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 1]
```diff
@@ -18,274 +18,113 @@
...
-     * <p>Actions performed at this step:</p>
-     * <ul>
-     * <li>Finds a checkbox within the context;</li>
-     * <li>Checks its state, if it's <b><i>not selected</i></b> then changes its state
-     * <li>Waits for page to load;</li>
-     * </ul>
-     */
-    @When("I check a checkbox")
-    public void checkCheckBox()
+    private CheckboxSteps(IBaseValidations baseValidations, IMouseActions mouseActions)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:41]: `public`?

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 2]
```diff
@@ -18,274 +18,113 @@
...
-     * then change it to <b><i>selected</i></b>;</li>
-     * <li>Waits for page to load;</li>
-     * </ul>
+     * Performs action on all the checkboxes found by locator
+     * @param checkBoxAction Actions to be performed (CHECK, UNCHECK)
+     * @param checkboxesLocator Locator to locate checkboxes
      */
-    @When("I check all the checkboxes")
-    public void checkAllCheckboxes()
+    @When("I $checkboxState all checkboxes located by `$checkboxesLocator`")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:52]: `$checkboxState` -> `$checkboxAction` (to keep naming consistency)

Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:52]: what about converting the step to composite one?

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 3]
```diff
@@ -18,274 +18,113 @@
...
+     * @param checkboxLocator Locator to locate checkboxes
      */
-    @When(value = "I $checkBoxAction a checkbox with the name '$checkBoxName'", priority = 1)
-    public void processCheckboxItem(CheckboxAction checkBoxAction, String checkBoxName)
+    @When(value = "I $checkBoxAction checkbox located by `$checkboxLocator`", priority = 1)
+    public void changeStateOfCheckbox(CheckboxAction checkBoxAction, SearchAttributes checkboxLocator)
     {
-        Checkbox checkBox = ifCheckboxExists(checkBoxName);
-        changeCheckboxState(checkBox, checkBoxAction.isSelected());
+        Checkbox checkBox = ifCheckboxExists(checkboxLocator);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:68]: `checkBox` -> `checkbox`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 4]
```diff
@@ -18,274 +18,113 @@
...
-     * </ul>
-     * @param checkBoxAction <b>CHECK</b> or <b>UNCHECK</b>
-     * @param checkBoxName Checkbox text (the text in the (<i><code>&lt;label&gt;</code></i>) tag
+     * Performs action on checkbox found by locator
+     * @param checkBoxAction Actions to be performed (CHECK, UNCHECK)
+     * @param checkboxLocator Locator to locate checkboxes
      */
-    @When(value = "I $checkBoxAction a checkbox with the name '$checkBoxName'", priority = 1)
-    public void processCheckboxItem(CheckboxAction checkBoxAction, String checkBoxName)
+    @When(value = "I $checkBoxAction checkbox located by `$checkboxLocator`", priority = 1)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:65]: is priority still required?

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 5]
```diff
@@ -18,274 +18,113 @@
...
-     * Attribute</i></a>
-     * @see <a href="https://www.w3schools.com/tags/default.asp"><i>HTML Element Reference</i></a>
+     * Verifies that checkbox by locator exists
+     * @param checkboxLocator Locator to locate checkboxes
+     * @return Found checkbox
      */
-    @When("I uncheck a checkbox with the attribute '$attributeType'='$attributeValue'")
-    public void uncheckCheckboxItem(String attributeType, String attributeValue)
+    @Then("checkbox located by `$searchAttributes` exists")
+    public Checkbox ifCheckboxExists(SearchAttributes checkboxLocator)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:78]: `ifCheckboxExists` -> `doesCheckboxExist` or `assertCheckboxExists`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 6]
```diff
@@ -18,274 +18,113 @@
...
-     * @param attributeValue A value of the attribute
-     * @return Web element - a <b>checkbox</b> that meets the requirements,
-     * <b> null</b> - if there are no expected elements.
+     * Verifies that checkbox located by locator is in an expected state.
+     * @param state State to check
+     * @param checkboxLocator Locator to locate checkbox
      */
-    @Then("a checkbox with the attribute '$attributeType'='$attributeValue' exists")
-    public Checkbox ifCheckboxWithAttributeExists(String attributeType, String attributeValue)
+    @Then("[$state] checkbox located by `$checkboxLocator` exists")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:99]:
```
@Then("$state checkbox located by `$checkboxLocator` exists")
```

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 7]
```diff
@@ -18,274 +18,113 @@
...
+    public void checkboxInStateExists(State state, SearchAttributes checkboxLocator)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:100]: `checkboxInStateExists` -> `doesCheckboxExist` or `assertCheckboxExists`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 8]
```diff
@@ -18,274 +18,113 @@
...
-        changeCheckboxState(attributeType, attributeValue, true);
-    }
-
-    private void changeCheckboxState(String attributeType, String attributeValue, boolean selected)
-    {
-        changeCheckboxState(getCheckboxXpathByAttributeAndValue(attributeType, attributeValue), selected);
-    }
-
-    private void changeCheckboxState(String xpath, boolean isSelected)
+    @Then("checkbox locted by `$checkboxLocator` does not exist")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:110]: I propose to replace such steps with already existing one:
```
Then number of elements found by `$locator` is $comparisonRule `$quantity`
```

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 9]
```diff
@@ -18,274 +18,113 @@
...
-     * @see <a href="https://www.w3schools.com/tags/att_input_checked.asp"><i>HTML </i>&lt;input&gt;<i> checked
-     * Attribute</i></a>
-     * @see <a href="https://www.w3schools.com/tags/default.asp"><i>HTML Element Reference</i></a>
+     * Verifies that checkbox by locator exists
+     * @param checkboxLocator Locator to locate checkboxes
+     * @return Found checkbox
      */
-    @When("I uncheck a checkbox with the attribute '$attributeType'='$attributeValue'")
-    public void uncheckCheckboxItem(String attributeType, String attributeValue)
+    @Then("checkbox located by `$searchAttributes` exists")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:77]: I propose to replace such steps with already existing one:
```
Then number of elements found by `$locator` is $comparisonRule `$quantity`
``

File: vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/SearchActionsTests.java [hunk 10]
```diff
@@ -70,7 +70,7 @@
     private DefaultSearch defaultSearch;
 
     @Mock
-    private CaseSensitiveTextSearch caseSensitivetextSearch;
+    private CaseSensitiveTextSearch caseSensitiveTextSearch;
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/SearchActionsTests.java:73]: I think it should go as a separate commit/PR

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java (hunk 2 above)
Comment (ikalinin1) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:52]: Could not be achieved without research overhead per checkbox

**Comment by ikalinin1** [comment]
Waits for #832  and #834

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 11]
```diff
@@ -18,274 +18,71 @@
...
-     * @param state A state value of the element
-     * (<i>Possible values:</i> <b>ENABLED, DISABLED, SELECTED, NOT_SELECTED, VISIBLE, NOT_VISIBLE</b>)
-     * @param checkboxName Checkbox text (the text in the (<i><code>&lt;label&gt;</code></i>) tag
+     * Performs action on all the checkboxes found by locator
+     * @param checkBoxAction Actions to be performed (CHECK, UNCHECK)
+     * @param checkboxesLocator Locator to locate checkboxes
      */
-    @Then("a [$state] checkbox with the name '$checkboxName' exists")
-    public void ifCheckboxExists(State state, String checkboxName)
+    @When("I $checkboxAction all checkboxes located by `$checkboxesLocator`")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:49]: `located by` -> `located`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 12]
```diff
@@ -18,274 +18,71 @@
...
-     * @see <a href="https://www.w3schools.com/tags/default.asp"><i>HTML Element Reference</i></a>
-     */
-    @When("I check a checkbox with the attribute '$attributeType'='$attributeValue'")
-    public void checkCheckboxItem(String attributeType, String attributeValue)
-    {
-        changeCheckboxState(attributeType, attributeValue, true);
-    }
-
-    private void changeCheckboxState(String attributeType, String attributeValue, boolean selected)
+    @When(value = "I $checkBoxAction checkbox located by `$checkboxLocator`")
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:62]: `located by` -> `located`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java [hunk 13]
```diff
@@ -18,274 +18,71 @@
...
+    public void changeStateOfCheckbox(CheckboxAction checkBoxAction, SearchAttributes checkboxLocator)
     {
-        changeCheckboxState(getCheckboxXpathByAttributeAndValue(attributeType, attributeValue), selected);
+        WebElement checkbox1 = baseValidations.assertIfElementExists(CHECKBOX, checkboxLocator);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/CheckboxSteps.java:65]: `checkbox1` -> `checkboxElement`

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]
